include/seqan3/alphabet/union_alphabet.hpp

```diff
@@ -31,237 +31,411 @@
+    //!\brief Return the letter's numeric value or rank in the alphabet.
     constexpr rank_type to_rank() const
     {
-        return value;
+        return _value;
     }
+    //!\}
 
-    //! assign from a character
+    /*!\name Write functions
+     * \{
+     */
+    //!\brief Assign from a character.
     constexpr union_alphabet & assign_char(char_type const c)
     {
-        auto sizes_it = sizes.begin();
-        auto sizes_it_end = sizes.end();
-        auto alphabets_it = alphabets.begin();
-        auto size_start = 0;
-
-        bool found = false;
-
-        while(sizes_it != sizes_it_end && !found)
-        {
-            std::visit([&](auto alphabet)
-            {
-                alphabet.assign_char(c);
-                if (alphabet.to_char() == c)
-                {
-                    value = size_start + alphabet.to_rank();
-                    found = true;
-                }
-            }, *alphabets_it);
-
-            size_start += *sizes_it;
-            sizes_it++;
-            alphabets_it++;
-        }
-
-        if (!found)
-        {
-            value = 0;
-        }
-
+        _value = char_to_value[c];
         return *this;
     }
 
-    //! assign from a numeric value
+    //!\brief Assign from a numeric value.
     constexpr union_alphabet & assign_rank(rank_type const i)
     {
         assert(i < value_size);
-        value = i;
+        _value = i;
         return *this;
     }
+    //!\}
 
-    //! @name comparison operators
-    //!@{
+    //!\name Comparison operators
+    //!\{
     constexpr bool operator==(union_alphabet const & rhs) const
     {
-        return value == rhs.value;
+        return _value == rhs._value;
     }
 
     constexpr bool operator!=(union_alphabet const & rhs) const
     {
-        return value != rhs.value;
+        return _value != rhs._value;
     }
 
     constexpr bool operator<(union_alphabet const & rhs) const
     {
-        return value < rhs.value;
+        return _value < rhs._value;
     }
 
     constexpr bool operator>(union_alphabet const & rhs) const
     {
-        return value > rhs.value;
+        return _value > rhs._value;
     }
 
     constexpr bool operator<=(union_alphabet const & rhs) const
     {
-        return value <= rhs.value;
+        return _value <= rhs._value;
     }
 
     constexpr bool operator>=(union_alphabet const & rhs) const
     {
-        return value >= rhs.value;
+        return _value >= rhs._value;
+    }
+    //!\}
+
+    //!\privatesection
+    //!\brief The data member.
+    rank_type _value;
+    //!\publicsection
+
+protected:
+    //!\privatesection
+
+    //!\brief Converts an object of one of the base alphabets into the internal representation
+    static constexpr rank_type from_base_(variant_type const & alphabet_v)
+    {
+        return std::visit([&](auto && alphabet) -> rank_type
+        {
+            return prefix_sum_sizes[alphabet_v.index()] + static_cast<rank_type>(alphabet.to_rank());
```

Comment:
Ah thank you, that's actually a good point. That also means that `::value_size` won't work.

I will add a test case and try to figure out how to fix that.